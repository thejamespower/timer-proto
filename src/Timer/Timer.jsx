import React from 'react'
import PropTypes from 'prop-types'
import Countdown from 'react-sexy-countdown'
import momentPropTypes from 'react-moment-proptypes'

const Timer = ({ timer }) => <Countdown date={timer.date.format()} />

Timer.propTypes = {
  timer: PropTypes.shape({
    id: PropTypes.string.isRequired,
    date: momentPropTypes.momentObj.isRequired,
  }).isRequired,
}

export default Timer
