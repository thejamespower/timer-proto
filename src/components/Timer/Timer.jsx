import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Countdown from 'react-countdown-now'
import moment from 'moment'
import Card from '@material-ui/core/Card'
import CardContent from '@material-ui/core/CardContent'
import Typography from '@material-ui/core/Typography'
import TimerDeleteButton from '../TimerDeleteButton/TimerDeleteButton'

class Timer extends Component {
  static propTypes = {
    timer: PropTypes.shape({
      id: PropTypes.string.isRequired,
      duration: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      active: PropTypes.bool.isRequired,
      timeToStart: PropTypes.string.isRequired,
      timeToStartInSeconds: PropTypes.number.isRequired,
      complete: PropTypes.bool.isRequired,
    }).isRequired,
    deleteTimer: PropTypes.func.isRequired,
    completeTimer: PropTypes.func.isRequired,
    superTimerActive: PropTypes.bool.isRequired,
  }

  constructor(props) {
    super(props)

    const { timer: { duration } } = this.props

    this.state = {
      duration,
    }

    this.handleComplete = this.handleComplete.bind(this)
  }

  shouldComponentUpdate(nextProps) {
    const { timer: { timeToStart: oldTimeToStart, active: oldActive } } = this.props
    const { timer: { timeToStart: newTimeToStart, active: newActive } } = nextProps
    return newActive !== oldActive || newTimeToStart !== oldTimeToStart
  }

  handleComplete() {
    const { completeTimer, timer: { id } } = this.props
    completeTimer(id)
  }

  render() {
    const {
      timer: {
        name, active, timeToStart, id, complete,
      },
      superTimerActive,
      deleteTimer,
    } = this.props

    const { duration } = this.state

    return (
      <Card>
        <CardContent>
          <Typography color="textSecondary" gutterBottom>Timer</Typography>
          <Typography>{name}</Typography>
          {active && <Countdown date={moment().add(moment.duration(duration))} onComplete={this.handleComplete} />}
          {!active && !complete && <p>{duration}, Time to start: {timeToStart}</p>}
          {complete && <p>Done!</p>}
        </CardContent>
        <TimerDeleteButton
          onClick={() => deleteTimer(id)}
          superTimerActive={superTimerActive}
          active={active}
          id={id}
          complete={complete}
        />
      </Card>
    )
  }
}

export default Timer
