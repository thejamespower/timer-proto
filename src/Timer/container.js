import { connect } from 'react-redux'

import Timer from './Timer'

const mapStateToProps = null

export default connect(
  mapStateToProps,
  null,
)(Timer)
