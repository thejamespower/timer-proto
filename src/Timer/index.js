import Timer from './container'

export default Timer
