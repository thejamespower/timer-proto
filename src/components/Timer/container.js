import { connect } from 'react-redux';

import { bindActionCreators } from 'redux';
import Timer from './Timer';
import { completeTimer } from '../../action-creators';

const mapStateToProps = state => ({
  superTimerActive: state.timers.superTimer.active,
});

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      completeTimer,
    },
    dispatch,
  );

export default connect(mapStateToProps, mapDispatchToProps)(Timer);
