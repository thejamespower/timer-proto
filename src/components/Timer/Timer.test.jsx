import React from 'react'
import renderer from 'react-test-renderer'
import { mount, shallow } from 'enzyme'
import Countdown from 'react-countdown-now'
import Timer from './Timer'

jest.mock('../TimerDeleteButton', () => () => 'TimerDeleteButton')

describe('Timer', () => {
  let props = {
    timer: {
      id: 'test',
      duration: '00:00:00',
      durationInSeconds: 0,
      name: '',
      active: false,
      timeToStart: '00:00:00',
      timeToStartInSeconds: 0,
      complete: false,
    },
    completeTimer: () => null,
    superTimerActive: false,
  }

  describe('render', () => {
    it('renders default state', () => {
      const component = renderer.create(
        <Timer {...props} />,
      )
      const tree = component.toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('renders active state', () => {
      props = {
        ...props,
        timer: {
          ...props.timer,
          active: true,
          duration: '00:00:01',
          durationInSeconds: 1,
        },
      }
      const component = renderer.create(
        <Timer {...props} />,
      )
      const tree = component.toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('renders complete state', () => {
      props = {
        ...props,
        timer: {
          ...props.timer,
          active: false,
          complete: true,
        },
      }
      const component = renderer.create(
        <Timer {...props} />,
      )
      const tree = component.toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('behaviour', () => {
    it('handles timer completion', () => {
      props = {
        ...props,
        timer: {
          ...props.timer,
          active: true,
        },
        completeTimer: jest.fn(),
      }

      const wrapper = mount(<Timer {...props} />)
      const countdown = wrapper.find(Countdown)
      countdown.prop('onComplete')()

      expect(props.completeTimer).toHaveBeenCalledWith(props.timer.id)
    })

    it('updates if active changes', () => {
      props = {
        ...props,
        timer: {
          ...props.timer,
          active: false,
        },
      }
      const wrapper = shallow(<Timer {...props} />)
      const nextProps = {
        ...props,
        timer: {
          ...props.timer,
          active: true,
        },
      }
      const shouldUpdate = wrapper.instance().shouldComponentUpdate(nextProps)
      expect(shouldUpdate).toBe(true)
    })

    it('updates if timeToStart changes', () => {
      const wrapper = shallow(<Timer {...props} />)
      const nextProps = {
        ...props,
        timer: {
          ...props.timer,
          timeToStart: '00:00:01',
        },
      }
      const shouldUpdate = wrapper.instance().shouldComponentUpdate(nextProps)
      expect(shouldUpdate).toBe(true)
    })
  })
})
