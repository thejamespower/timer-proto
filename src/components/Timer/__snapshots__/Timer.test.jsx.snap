// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer render renders active state 1`] = `
<div
  className="MuiPaper-root-2 MuiPaper-elevation1-5 MuiPaper-rounded-3 MuiCard-root-1"
>
  <div
    className="MuiCardContent-root-29"
  >
    <p
      className="MuiTypography-root-30 MuiTypography-body1-39 MuiTypography-colorTextSecondary-63 MuiTypography-gutterBottom-57"
    >
      Timer
    </p>
    <p
      className="MuiTypography-root-30 MuiTypography-body1-39"
    >
      
    </p>
    <span>
      0
      :
      00
      :
      00
      :
      01
    </span>
  </div>
  TimerDeleteButton
</div>
`;

exports[`Timer render renders complete state 1`] = `
<div
  className="MuiPaper-root-2 MuiPaper-elevation1-5 MuiPaper-rounded-3 MuiCard-root-1"
>
  <div
    className="MuiCardContent-root-29"
  >
    <p
      className="MuiTypography-root-30 MuiTypography-body1-39 MuiTypography-colorTextSecondary-63 MuiTypography-gutterBottom-57"
    >
      Timer
    </p>
    <p
      className="MuiTypography-root-30 MuiTypography-body1-39"
    >
      
    </p>
    <p>
      Done!
    </p>
  </div>
  TimerDeleteButton
</div>
`;

exports[`Timer render renders default state 1`] = `
<div
  className="MuiPaper-root-2 MuiPaper-elevation1-5 MuiPaper-rounded-3 MuiCard-root-1"
>
  <div
    className="MuiCardContent-root-29"
  >
    <p
      className="MuiTypography-root-30 MuiTypography-body1-39 MuiTypography-colorTextSecondary-63 MuiTypography-gutterBottom-57"
    >
      Timer
    </p>
    <p
      className="MuiTypography-root-30 MuiTypography-body1-39"
    >
      
    </p>
    <p>
      00:00:00
      , Time to start: 
      00:00:00
    </p>
  </div>
  TimerDeleteButton
</div>
`;
